File: database_connection/task_list/task_list_node/service/TaskService.js
const TaskDto = require("../dto/TaskDto");
const Task = require("../model/Task");
const TaskRepository = require("../repository/TaskRepository");

/**
 * @class TaskService
 */
class TaskService {
  /**
   * @param {TaskRepository} taskRepository
   */
  constructor(taskRepository) {
    this.taskRepository = taskRepository;
  }

  /**
   * @param {TaskDto} taskDto
   */
  async addNewTask(taskDto) {
    const task = Task.fromDto(taskDto);
    await this.taskRepository.insert(task);
  }

  /**
   * @param {number} id
   * @param {TaskDto} taskDto
   */
  async updateTask(id, taskDto) {
    const task = await this.retrieveTaskById(id);
    task.updateData(taskDto);
    await this.taskRepository.update(id, task);
  }

  /**
   * @param {number} id
   */
  async deleteTaskById(id) {
    await this.retrieveTaskById(id); // just to verify if task exists
    await this.taskRepository.deleteById(id);
  }

  /**
   * @param {number} id
   * @returns {Task}
   */
  async retrieveTaskById(id) {
    return await this.taskRepository.findById(id);
  }

  /**
   * @returns {Task[]}
   */
  async retrieveAllTasks() {
    return await this.taskRepository.findAll();
  }
}

module.exports = TaskService;

File: database_connection/task_list/task_list_python_venv/service/task_service.py
from dto.task_dto import TaskDto
from model.task import Task
from repository.task_repository import TaskRepository


class TaskService:
  def __init__(self, task_repository: TaskRepository) -> None:
    self.task_repository = task_repository

  def add_new_task(self, task_dto: TaskDto):
    task = Task.from_dto(task_dto)
    self.task_repository.insert(task)

  def update_task(self, id: int, task_dto: TaskDto):
    task = self.retrieve_task_by_id(id)
    task.update_data(task_dto)
    self.task_repository.update(id, task)

  def delete_task_by_id(self, id: int):
    self.retrieve_task_by_id(id)
    self.task_repository.delete_by_id(id)

  def retrieve_task_by_id(self, id: int) -> Task:
    return self.task_repository.find_by_id(id)

  def retrieve_all_tasks(self) -> list[Task]:
    return self.task_repository.find_all()
